crypto/pem/pem_lib.c

```diff
@@ -661,177 +694,288 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
 }
 #endif
 
-int PEM_read_bio(BIO *bp, char **name, char **header, unsigned char **data,
-                 long *len)
+/* Some helpers for PEM_read_bio_flags(). */
+
+/* TODO Presumably this should be a table lookup, but are these the right ranges? */
+static int isb64(char c)
```

Comment:
Not sure why you would want a table; in all codesets those ranges are contiguous.  The issues come when you assume the relative placements of upper vs lower case, or digits vs letters, or where other graphic (e.g. + and /) appear relative to other characters.

Not that this is particularly performance sensitive, but a table (256 bytes? bigger?) would require memory and data cache; the range tests are compact (minimum instruction and no data cache impact) and cover the cases.  I don't think a table adds clarity to the code.  The ranges capture it.  This isn't crypto, so no reason to care about constant time.  So, I'd stick with what you have.

You might return -1 for invalid, and the encoded value for OK (e.g. 0 for 'A' & 26 for 'a'); that would let the caller do something like `if( (nibble = isb64(in)) < 0 ) return error (or ignore); val |= nibble << state;`  It's easy to do: `if( c >= 'a' && c <= 'z' ) return (c - 'a') + 26;` etc.

You might want a state machine somewhere.; that might work as a (small) table.  (I haven't reviewed the whole commit, and don't plan to.  But noticed the questions.)

The ranges look right to me, assuming you handle pad ('=') elsewhere.  Ref: [https://tools.ietf.org/html/rfc4648#page-5](rfc4648), which  has a table.

My inclination for default is to be lax; interpret B64 chars, ignore everything else.  Pad only at the end, and only the permissible cases (0,1,2).  There's a case for exception if non-B64 other than whitespace (which is sp,ht,lf,cr,ff, & often vt; sometimes NUL) or proper padding.  The EAY line length-sensitive weirdness is at best a desperate fallback for compatibility; should not be a default.  (Even though it is old behavior, it causes reasonable input to be rejected.)

I'm assuming that the apps / consumers will be updated as needed to resolve the external manifestations. Hopefully a consistent option: `-pemfmt [obsolete|strict]` with default = `lax` ?

Hope this helps.  Thanks for working on it.